XLSX export formatting

noahmoss opened the pull request:
This PR adds number, currency, and datetime formatting for XLSX exports. Sorry for the mega PR; hopefully it's easy enough to figure out what's going on from the code & comments.

The high level approach is:
1. An ordered & filtered list of column metadata, and viz settings for a question, are being passed into `begin!` and `write-row!` functions of the `streaming-results-writer` in `metabase.query-processor.streaming.xlsx`. This was added in the previous PR (#16972).

2. Column titles are generated in `begin!` and written to the export. `::mb.viz/column-title` in the viz settings takes precedence over `:name` or `:display_name` from the col metadata. A currency is also added to the column title if necessary.

3. The first time `write-row!` is called, it calls `cell-style-delays`, which generates format string corresponding to each columns' format settings. It then creates a map with keys being the ids (or names, for native queries) of columns, and values being delays which, when de-referenced, add the format string to the workbook. (This was the existing approach, but just for default datetime formats).

4. This map is passed into the `add-row!` multimethod which ultimately calls `set-cell!` to write each value in the row to the workbook. `set-cell!` fetches the appropriate format string from the map and adds it to the workbook as well.

5. `cell-style-delays` is memoized, so for subsequent rows, `write-row!` calls it and gets a cached value of the map with format string delays.

There are a number of edge cases that are handled by the code but I tried to leave comments explaining them. Let me know if anything is confusing or if there's any edge cases I haven't considered.

noahmoss committed:
initial work

noahmoss committed:
memoize column-style-delays and fix warnings

noahmoss committed:
all number formatting aside from currencies

noahmoss committed:
currency formatting

noahmoss committed:
wip datetime formatting

noahmoss committed:
call createDataFormat only once

noahmoss committed:
Merge branch 'master' into export-formatting

noahmoss committed:
WIP column titles

noahmoss committed:
determine col order by either name or id

noahmoss committed:
don't remove empty clauses in viz settings

noahmoss committed:
rename variables

noahmoss committed:
time format fixes

noahmoss committed:
fix currency edge case

noahmoss committed:
include currency in header

noahmoss committed:
small fixes

noahmoss committed:
unit tests for format string generation

noahmoss committed:
reorganize xlsx code

noahmoss committed:
Merge branch 'master' into export-formatting

noahmoss committed:
export streaming tests

noahmoss committed:
namespace cleanup

noahmoss committed:
test refactor

noahmoss committed:
column correlation test

noahmoss committed:
misc fixes and an extra test

howonlee commented on src/metabase/query_processor/streaming/xlsx.clj:
```diff
@@ -86,25 +284,44 @@
-  (spreadsheet/set-cell! cell (t/offset-date-time t)))
+(defmethod set-cell! String
+  [^Cell cell value _]
+  (when (= (.getCellType cell) CellType/FORMULA)
```
I don't know if this prevents xlsx injection: I will go find out, lol

howonlee replied:
I can get a formula box that gets a shell by noodling around a bit in repl, I am going to try to repro going through the whole thing

noahmoss replied:
just curious, how did you test this?

howonlee replied:
if you can get any formula to run upon xlsx open, you can get the excessively "fun" formula to run, including stuff that can exfiltrate data or get shell access

https://labs.bishopfox.com/tech-blog/2018/06/server-side-spreadsheet-injections

docjure seems to take care of it for you luckily enough

howonlee replied:
actually now that i think about it im going to try it on the csvs

mazameli commented:
I exported to XLSX and opened the file in Numbers. Some things I noticed:

👍🏼 
- My custom column names were respected
- Multiplying my Price column by 100 was respected, as well as the dollar symbol

👎🏼 
- My Ordered At column looks completely wrong. It says e.g. `D.2.YYYY` instead of the expected `11.2.2019`
- My TOTAL column has way more decimal places than in my question, maybe because I didn't apply any custom formatting to it?
- (Known limitation) My Price column didn't carry over my custom separators.


![image](https://user-images.githubusercontent.com/2223916/128389930-fb8c7a5b-326b-4a29-9680-4b3418a44f98.png)

noahmoss commented:
>     * My Ordered At column looks completely wrong. It says e.g. `D.2.YYYY` instead of the expected `11.2.2019`

Hmmmm, this seems to be a discrepancy between Excel and Numbers. I've been testing in Excel and this example works fine. I'll dig more into it.

>     * My TOTAL column has way more decimal places than in my question, maybe because I didn't apply any custom formatting to it?

This is also a known limitation...I'm currently just using Excel's default number formatting if no custom formatting is set. Would it be better to default to two decimal points? Then integers would show up as `3.00`, etc. There's unfortunately no way (to my knowledge) to cap it at a certain number of decimal points, but hide trailing zeros.

noahmoss committed:
include viz-settings for embedded questions

noahmoss committed:
lowercase datetime format strings so that they work in Numbers

noahmoss committed:
better default number formatting

mazameli commented:
Looking pretty dang great now, man. 👏🏼 

![image](https://user-images.githubusercontent.com/2223916/128436573-01d39e7d-e0b1-445c-92be-27865d40d166.png)

noahmoss committed:
fix test failures

noahmoss committed:
fix edge case for currency symbols in headers

flamber commented:
This looks great - testing commit `3397366` - just did a quick test, a couple of things that I noticed:

1. Unless you have defined column formatting (either in Data Model or viz), then it seems to use Metabase defaults instead of global defaults (Admin > Settings > Localization) - I've written about the 5 different levels of formatting #11394
![image](https://user-images.githubusercontent.com/1447303/128510129-016a0665-5d18-42ab-ba07-5861a61fa2da.png)
2. It's applying thousand separators to PK/FK columns
![image](https://user-images.githubusercontent.com/1447303/128510213-633a8caf-5d1e-4da1-9a4b-a989b3e3e3c7.png)

noahmoss committed:
fix formatting for pks and fks

noahmoss committed:
incorporate global currency and datetime format settings

noahmoss committed:
clean ns

noahmoss committed:
fix lint errors

noahmoss committed:
Merge branch 'master' into export-formatting

noahmoss committed:
fix edge case for global datetime format settings

noahmoss commented:
@flamber I believe I've fixed both the issues you identified. Let me know if that's not the case.

flamber commented:
@noahmoss Awesome. Did some more extensive testing of `89456a7`. Some of this might not be specific to this PR.

1. It seems to telling PK/FK to be strings (perhaps because of PR 13107) - if that's the only way to avoid the separator formatting, then I prefer to have it as numeric with the separator formatting.
2. When I download unsaved question, then it's different from saved question
![image](https://user-images.githubusercontent.com/1447303/128597836-e5351618-177e-4346-af81-6a5428e5557b.png)
Saved question - includes the remapped column (though incorrect column title, but there are some issues related to that)
![image](https://user-images.githubusercontent.com/1447303/128597469-6c3f1920-5e87-49f6-8486-e4b68bb74ae6.png)
Unsaved question (just swapped Tax and Total column) - no remapped column and PK/FK are now numeric
![image](https://user-images.githubusercontent.com/1447303/128597486-a7dabba5-62e1-4496-a5b4-998a472250ff.png)
3. If I download from Public Sharing, then it seems like it's applying global formatting (besides timestamps, which has it's own issue), but not Data Model nor viz formatting. Embedded download is correct (besides timestamps). I didn't test Pulse attachment.
But can we somehow create a unification of exports, so we always produce the same results (of course according to permissions etc)? We have had a lot of regressions because of the difference in exports.
![image](https://user-images.githubusercontent.com/1447303/128597532-bba79360-9572-43cc-af54-530bae9f8e86.png)
4. I can see that XLSX creation takes about twice as long compared to 0.40.2, when I do various combinations of Sample Dataset. That is to be expected I guess, just wanted to note it. It might become problematic if someone is exporting 1 million rows from a 20+ column table. Could we do a bit of testing on memory usage compared to 0.40.2?

But during my testing I ended up with a stacktrace and an issue I have been trying to reproduce for two weeks, which should have been fixed on master PR 17279 (included in 0.40.2). My current thought is that it's caused by `table.columns`, but I don't think it's related to this PR, since I can see you just included master - though, not sure why it triggered a stacktrace on export.
I'm still investigating this, but keeping it separate from this PR.

noahmoss committed:
faster version of rounds-to-int?

noahmoss committed:
fix public formatting, and keep pks/fks as ints

noahmoss committed:
fix `export-column-order` when cols have been remapped

noahmoss committed:
fix viz settings tests

noahmoss commented:
@flamber 

1. This should be fixed now.

2. I've fixed a bug in the column ordering logic which, I think, should make remapped columns work in all cases now. What exactly did you mean by "incorrect column title"?

3. Public & embedded downloads should now have all formatting applied. Pulse attachments will require a bit more refactor so I'll leave it out of this PR to avoid making it even bigger. As of this PR, pulse attachments will have global formatting applied only.

4. Good catch, the previous approach I used for default number formatting was a CPU hot spot. I changed it to a much faster approach; see if that helps. (There's some unavoidable overhead for formatting but it definitely shouldn't be a 2x slowdown.)

5. I'm using `table.columns` to order columns for the export, so if it has duplicate columns that might have messed up the logic somewhere and caused a stacktrace. Is there a way I can repro this?

dpsutton commented on src/metabase/api/card.clj:
```diff
@@ -629,8 +629,10 @@
-                  (update :middleware merge {:js-int-to-string?      true
-                                             :ignore-cached-results? ignore_cache}))
+                  (update :middleware (fn [middleware]
+                                        (merge
```
why the swap of the merge order here?

noahmoss replied:
The `middleware` argument should take precedence since we need `js-int-to-string?` to be false for exports, so that pks/fks aren't converted to strings.

In the current code, the passed-in `middleware` is always overwritten by the hard-coded values here.

flamber commented:
@noahmoss Nice! Tested `8758b8d`
1. Perfect
2. It's now correctly including the remapped field in the correct position on unsaved questions, but saved/public/embedded does not include the remapped field at all anymore.
Unsaved question:
![image](https://user-images.githubusercontent.com/1447303/128745190-339804a8-674e-47b0-94df-bda34b51ca3a.png)
Saved/public/embed question:
![image](https://user-images.githubusercontent.com/1447303/128745285-da85566f-f4fc-4aa9-8ba8-55e18d585092.png)
3. Formatting is consistent across all (saved, unsaved, public, embed)
4. Definitely much faster now - across a few tests I would say it's only 10% slower than 0.40.2, which is totally within expectation.
5. I was messing around with an explicit join (Orders joining in Products), where I then did some column modifications and then I ended up with it constantly showing the Run-overlay until I removed the joined table. I then saved the question, and then tried exporting it. I would guess it might be caused by orphaned table.columns (or was it column_settings), which has been an issue on Pulse viz previously.
<details><summary>5 stacktrace</summary>

```
2021-08-09 19:12:53,701 ERROR middleware.catch-exceptions :: Error processing query: null
{:database_id 4,
 :started_at #t "2021-08-09T19:12:52.619679+02:00[Europe/Copenhagen]",
 :via
 [{:status :failed,
   :class clojure.lang.ExceptionInfo,
   :error "Error reducing result rows",
   :stacktrace
   ["--> query_processor.context.default$default_reducef$fn__42144.invoke(default.clj:60)"
    "query_processor.context.default$default_reducef.invokeStatic(default.clj:57)"
    "query_processor.context.default$default_reducef.invoke(default.clj:48)"
    "query_processor.context$reducef.invokeStatic(context.clj:69)"
    "query_processor.context$reducef.invoke(context.clj:62)"
    "query_processor.context.default$default_runf$respond_STAR___42148.invoke(default.clj:69)"
    "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:485)"
    "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:463)"
    "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:472)"
    "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:463)"
    "driver.sql_jdbc$fn__84049.invokeStatic(sql_jdbc.clj:54)"
    "driver.sql_jdbc$fn__84049.invoke(sql_jdbc.clj:52)"
    "driver.h2$fn__81265.invokeStatic(h2.clj:83)"
    "driver.h2$fn__81265.invoke(h2.clj:80)"
    "query_processor.context$executef.invokeStatic(context.clj:59)"
    "query_processor.context$executef.invoke(context.clj:48)"
    "query_processor.context.default$default_runf.invokeStatic(default.clj:68)"
    "query_processor.context.default$default_runf.invoke(default.clj:66)"
    "query_processor.context$runf.invokeStatic(context.clj:45)"
    "query_processor.context$runf.invoke(context.clj:39)"
    "query_processor.reducible$pivot.invokeStatic(reducible.clj:34)"
    "query_processor.reducible$pivot.invoke(reducible.clj:31)"
    "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__48748.invoke(mbql_to_native.clj:25)"
    "query_processor.middleware.check_features$check_features$fn__49237.invoke(check_features.clj:39)"
    "query_processor.middleware.limit$limit$fn__47159.invoke(limit.clj:37)"
    "query_processor.middleware.cache$maybe_return_cached_results$fn__49611.invoke(cache.clj:204)"
    "query_processor.middleware.optimize_temporal_filters$optimize_temporal_filters$fn__50603.invoke(optimize_temporal_filters.clj:204)"
    "query_processor.middleware.validate_temporal_bucketing$validate_temporal_bucketing$fn__50647.invoke(validate_temporal_bucketing.clj:50)"
    "query_processor.middleware.auto_parse_filter_values$auto_parse_filter_values$fn__48807.invoke(auto_parse_filter_values.clj:43)"
    "query_processor.middleware.wrap_value_literals$wrap_value_literals$fn__38981.invoke(wrap_value_literals.clj:161)"
    "query_processor.middleware.annotate$add_column_info$fn__43394.invoke(annotate.clj:608)"
    "query_processor.middleware.permissions$check_query_permissions$fn__45314.invoke(permissions.clj:81)"
    "query_processor.middleware.pre_alias_aggregations$pre_alias_aggregations$fn__49786.invoke(pre_alias_aggregations.clj:40)"
    "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__45712.invoke(cumulative_aggregations.clj:60)"
    "query_processor.middleware.visualization_settings$update_viz_settings$fn__45650.invoke(visualization_settings.clj:56)"
    "query_processor.middleware.resolve_joined_fields$resolve_joined_fields$fn__46757.invoke(resolve_joined_fields.clj:102)"
    "query_processor.middleware.resolve_joins$resolve_joins$fn__50373.invoke(resolve_joins.clj:171)"
    "query_processor.middleware.add_implicit_joins$add_implicit_joins$fn__49908.invoke(add_implicit_joins.clj:190)"
    "query_processor.middleware.large_int_id$convert_id_to_string$fn__46776.invoke(large_int_id.clj:59)"
    "query_processor.middleware.format_rows$format_rows$fn__49960.invoke(format_rows.clj:74)"
    "query_processor.middleware.add_default_temporal_unit$add_default_temporal_unit$fn__46148.invoke(add_default_temporal_unit.clj:23)"
    "query_processor.middleware.desugar$desugar$fn__45623.invoke(desugar.clj:21)"
    "query_processor.middleware.binning$update_binning_strategy$fn__38881.invoke(binning.clj:229)"
    "query_processor.middleware.resolve_fields$resolve_fields$fn__44963.invoke(resolve_fields.clj:34)"
    "query_processor.middleware.add_dimension_projections$add_remapping$fn__49162.invoke(add_dimension_projections.clj:314)"
    "query_processor.middleware.add_implicit_clauses$add_implicit_clauses$fn__49460.invoke(add_implicit_clauses.clj:147)"
    "query_processor.middleware.upgrade_field_literals$upgrade_field_literals$fn__47144.invoke(upgrade_field_literals.clj:40)"
    "query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__46525.invoke(add_source_metadata.clj:123)"
    "query_processor.middleware.reconcile_breakout_and_order_by_bucketing$reconcile_breakout_and_order_by_bucketing$fn__49710.invoke(reconcile_breakout_and_order_by_bucketing.clj:100)"
    "query_processor.middleware.auto_bucket_datetimes$auto_bucket_datetimes$fn__48068.invoke(auto_bucket_datetimes.clj:147)"
    "query_processor.middleware.resolve_source_table$resolve_source_tables$fn__44944.invoke(resolve_source_table.clj:45)"
    "query_processor.middleware.parameters$substitute_parameters$fn__47746.invoke(parameters.clj:111)"
    "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__45017.invoke(resolve_referenced.clj:79)"
    "query_processor.middleware.expand_macros$expand_macros$fn__51031.invoke(expand_macros.clj:184)"
    "query_processor.middleware.add_timezone_info$add_timezone_info$fn__47525.invoke(add_timezone_info.clj:15)"
    "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__49975.invoke(splice_params_in_response.clj:32)"
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__49471$fn__49475.invoke(resolve_database_and_driver.clj:31)"
    "driver$do_with_driver.invokeStatic(driver.clj:60)"
    "driver$do_with_driver.invoke(driver.clj:56)"
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__49471.invoke(resolve_database_and_driver.clj:25)"
    "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__45561.invoke(fetch_source_query.clj:274)"
    "query_processor.middleware.store$initialize_store$fn__45749$fn__45750.invoke(store.clj:11)"
    "query_processor.store$do_with_store.invokeStatic(store.clj:44)"
    "query_processor.store$do_with_store.invoke(store.clj:38)"
    "query_processor.middleware.store$initialize_store$fn__45749.invoke(store.clj:10)"
    "query_processor.middleware.validate$validate_query$fn__49758.invoke(validate.clj:10)"
    "query_processor.middleware.normalize_query$normalize$fn__49765.invoke(normalize_query.clj:22)"
    "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__47474.invoke(add_rows_truncated.clj:35)"
    "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__48733.invoke(results_metadata.clj:143)"
    "query_processor.middleware.constraints$add_default_userland_constraints$fn__47492.invoke(constraints.clj:42)"
    "query_processor.middleware.process_userland_query$process_userland_query$fn__49746.invoke(process_userland_query.clj:134)"
    "query_processor.middleware.catch_exceptions$catch_exceptions$fn__50056.invoke(catch_exceptions.clj:173)"
    "query_processor.reducible$async_qp$qp_STAR___42232$thunk__42233.invoke(reducible.clj:103)"
    "query_processor.reducible$async_qp$qp_STAR___42232$fn__42235.invoke(reducible.clj:108)"],
   :error_type :qp,
   :ex-data {:type :qp}}],
 :error_type :qp,
 :json_query
 {:constraints nil,
  :type :query,
  :middleware
  {:js-int-to-string? false,
   :ignore-cached-results? true,
   :process-viz-settings? true,
   :skip-results-metadata? true,
   :format-rows? false},
  :database 4,
  :query {:source-table 11},
  :parameters nil,
  :async? true,
  :cache-ttl nil},
 :native
 {:query
  "SELECT \"PUBLIC\".\"ORDERS\".\"ID\" AS \"ID\", \"PUBLIC\".\"ORDERS\".\"USER_ID\" AS \"USER_ID\", \"PUBLIC\".\"ORDERS\".\"PRODUCT_ID\" AS \"PRODUCT_ID\", \"PUBLIC\".\"ORDERS\".\"SUBTOTAL\" AS \"SUBTOTAL\", \"PUBLIC\".\"ORDERS\".\"TAX\" AS \"TAX\", \"PUBLIC\".\"ORDERS\".\"TOTAL\" AS \"TOTAL\", \"PUBLIC\".\"ORDERS\".\"DISCOUNT\" AS \"DISCOUNT\", \"PUBLIC\".\"ORDERS\".\"CREATED_AT\" AS \"CREATED_AT\", \"PUBLIC\".\"ORDERS\".\"QUANTITY\" AS \"QUANTITY\", \"PRODUCTS__via__PRODUCT_ID\".\"TITLE\" AS \"PRODUCTS__via__PRODUCT_ID__TITLE\" FROM \"PUBLIC\".\"ORDERS\" LEFT JOIN \"PUBLIC\".\"PRODUCTS\" \"PRODUCTS__via__PRODUCT_ID\" ON \"PUBLIC\".\"ORDERS\".\"PRODUCT_ID\" = \"PRODUCTS__via__PRODUCT_ID\".\"ID\" LIMIT 1048575",
  :params nil},
 :status :failed,
 :class java.lang.IllegalArgumentException,
 :stacktrace
 ["clojure.lang.APersistentVector.invoke(APersistentVector.java:294)"
  "--> query_processor.streaming$streaming_rff$fn__37720$iter__37722__37726$fn__37727$fn__37728.invoke(streaming.clj:66)"
  "query_processor.streaming$streaming_rff$fn__37720$iter__37722__37726$fn__37727.invoke(streaming.clj:66)"
  "query_processor.streaming.xlsx$column_titles$iter__37648__37652$fn__37653.invoke(xlsx.clj:398)"
  "query_processor.streaming.xlsx$fn$reify__37679.begin_BANG_(xlsx.clj:420)"
  "query_processor.streaming$streaming_rff$fn__37720$fn__37735.invoke(streaming.clj:73)"
  "query_processor.middleware.process_userland_query$add_and_save_execution_info_xform_BANG_$execution_info_rf_STAR___49737.invoke(process_userland_query.clj:82)"
  "query_processor.middleware.add_rows_truncated$add_rows_truncated_xform$fn__47469.invoke(add_rows_truncated.clj:17)"
  "query_processor.middleware.limit$limit_xform$fn__47156.invoke(limit.clj:18)"
  "query_processor.context.default$default_reducef$fn__42144.invoke(default.clj:58)"
  "query_processor.context.default$default_reducef.invokeStatic(default.clj:57)"
  "query_processor.context.default$default_reducef.invoke(default.clj:48)"
  "query_processor.context$reducef.invokeStatic(context.clj:69)"
  "query_processor.context$reducef.invoke(context.clj:62)"
  "query_processor.context.default$default_runf$respond_STAR___42148.invoke(default.clj:69)"
  "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:485)"
  "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:463)"
  "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:472)"
  "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:463)"
  "driver.sql_jdbc$fn__84049.invokeStatic(sql_jdbc.clj:54)"
  "driver.sql_jdbc$fn__84049.invoke(sql_jdbc.clj:52)"
  "driver.h2$fn__81265.invokeStatic(h2.clj:83)"
  "driver.h2$fn__81265.invoke(h2.clj:80)"
  "query_processor.context$executef.invokeStatic(context.clj:59)"
  "query_processor.context$executef.invoke(context.clj:48)"
  "query_processor.context.default$default_runf.invokeStatic(default.clj:68)"
  "query_processor.context.default$default_runf.invoke(default.clj:66)"
  "query_processor.context$runf.invokeStatic(context.clj:45)"
  "query_processor.context$runf.invoke(context.clj:39)"
  "query_processor.reducible$pivot.invokeStatic(reducible.clj:34)"
  "query_processor.reducible$pivot.invoke(reducible.clj:31)"
  "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__48748.invoke(mbql_to_native.clj:25)"
  "query_processor.middleware.check_features$check_features$fn__49237.invoke(check_features.clj:39)"
  "query_processor.middleware.limit$limit$fn__47159.invoke(limit.clj:37)"
  "query_processor.middleware.cache$maybe_return_cached_results$fn__49611.invoke(cache.clj:204)"
  "query_processor.middleware.optimize_temporal_filters$optimize_temporal_filters$fn__50603.invoke(optimize_temporal_filters.clj:204)"
  "query_processor.middleware.validate_temporal_bucketing$validate_temporal_bucketing$fn__50647.invoke(validate_temporal_bucketing.clj:50)"
  "query_processor.middleware.auto_parse_filter_values$auto_parse_filter_values$fn__48807.invoke(auto_parse_filter_values.clj:43)"
  "query_processor.middleware.wrap_value_literals$wrap_value_literals$fn__38981.invoke(wrap_value_literals.clj:161)"
  "query_processor.middleware.annotate$add_column_info$fn__43394.invoke(annotate.clj:608)"
  "query_processor.middleware.permissions$check_query_permissions$fn__45314.invoke(permissions.clj:81)"
  "query_processor.middleware.pre_alias_aggregations$pre_alias_aggregations$fn__49786.invoke(pre_alias_aggregations.clj:40)"
  "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__45712.invoke(cumulative_aggregations.clj:60)"
  "query_processor.middleware.visualization_settings$update_viz_settings$fn__45650.invoke(visualization_settings.clj:56)"
  "query_processor.middleware.resolve_joined_fields$resolve_joined_fields$fn__46757.invoke(resolve_joined_fields.clj:102)"
  "query_processor.middleware.resolve_joins$resolve_joins$fn__50373.invoke(resolve_joins.clj:171)"
  "query_processor.middleware.add_implicit_joins$add_implicit_joins$fn__49908.invoke(add_implicit_joins.clj:190)"
  "query_processor.middleware.large_int_id$convert_id_to_string$fn__46776.invoke(large_int_id.clj:59)"
  "query_processor.middleware.format_rows$format_rows$fn__49960.invoke(format_rows.clj:74)"
  "query_processor.middleware.add_default_temporal_unit$add_default_temporal_unit$fn__46148.invoke(add_default_temporal_unit.clj:23)"
  "query_processor.middleware.desugar$desugar$fn__45623.invoke(desugar.clj:21)"
  "query_processor.middleware.binning$update_binning_strategy$fn__38881.invoke(binning.clj:229)"
  "query_processor.middleware.resolve_fields$resolve_fields$fn__44963.invoke(resolve_fields.clj:34)"
  "query_processor.middleware.add_dimension_projections$add_remapping$fn__49162.invoke(add_dimension_projections.clj:314)"
  "query_processor.middleware.add_implicit_clauses$add_implicit_clauses$fn__49460.invoke(add_implicit_clauses.clj:147)"
  "query_processor.middleware.upgrade_field_literals$upgrade_field_literals$fn__47144.invoke(upgrade_field_literals.clj:40)"
  "query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__46525.invoke(add_source_metadata.clj:123)"
  "query_processor.middleware.reconcile_breakout_and_order_by_bucketing$reconcile_breakout_and_order_by_bucketing$fn__49710.invoke(reconcile_breakout_and_order_by_bucketing.clj:100)"
  "query_processor.middleware.auto_bucket_datetimes$auto_bucket_datetimes$fn__48068.invoke(auto_bucket_datetimes.clj:147)"
  "query_processor.middleware.resolve_source_table$resolve_source_tables$fn__44944.invoke(resolve_source_table.clj:45)"
  "query_processor.middleware.parameters$substitute_parameters$fn__47746.invoke(parameters.clj:111)"
  "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__45017.invoke(resolve_referenced.clj:79)"
  "query_processor.middleware.expand_macros$expand_macros$fn__51031.invoke(expand_macros.clj:184)"
  "query_processor.middleware.add_timezone_info$add_timezone_info$fn__47525.invoke(add_timezone_info.clj:15)"
  "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__49975.invoke(splice_params_in_response.clj:32)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__49471$fn__49475.invoke(resolve_database_and_driver.clj:31)"
  "driver$do_with_driver.invokeStatic(driver.clj:60)"
  "driver$do_with_driver.invoke(driver.clj:56)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__49471.invoke(resolve_database_and_driver.clj:25)"
  "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__45561.invoke(fetch_source_query.clj:274)"
  "query_processor.middleware.store$initialize_store$fn__45749$fn__45750.invoke(store.clj:11)"
  "query_processor.store$do_with_store.invokeStatic(store.clj:44)"
  "query_processor.store$do_with_store.invoke(store.clj:38)"
  "query_processor.middleware.store$initialize_store$fn__45749.invoke(store.clj:10)"
  "query_processor.middleware.validate$validate_query$fn__49758.invoke(validate.clj:10)"
  "query_processor.middleware.normalize_query$normalize$fn__49765.invoke(normalize_query.clj:22)"
  "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__47474.invoke(add_rows_truncated.clj:35)"
  "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__48733.invoke(results_metadata.clj:143)"
  "query_processor.middleware.constraints$add_default_userland_constraints$fn__47492.invoke(constraints.clj:42)"
  "query_processor.middleware.process_userland_query$process_userland_query$fn__49746.invoke(process_userland_query.clj:134)"
  "query_processor.middleware.catch_exceptions$catch_exceptions$fn__50056.invoke(catch_exceptions.clj:173)"
  "query_processor.reducible$async_qp$qp_STAR___42232$thunk__42233.invoke(reducible.clj:103)"
  "query_processor.reducible$async_qp$qp_STAR___42232$fn__42235.invoke(reducible.clj:108)"],
 :context :xlsx-download,
 :error "Key must be integer",
 :row_count 0,
 :running_time 0,
 :preprocessed
 {:type :query,
  :middleware
  {:js-int-to-string? false,
   :ignore-cached-results? true,
   :process-viz-settings? true,
   :skip-results-metadata? true,
   :format-rows? false},
  :info
  {:executed-by 1,
   :context :xlsx-download,
   :card-id 723,
   :query-hash
   [89, -74, 49, -78, 114, 102, -34, 80, 95, 86, -31, 12, 61, 41, 114, 101, -9, -64, -90, -3, -95, -6, 97, -120, -51,
    61, -35, 11, 53, -105, 117, -71]},
  :database 4,
  :query
  {:source-table 11,
   :fields
   [[:field 83 nil]
    [:field 80 nil]
    [:field 82 nil]
    [:field 84 nil]
    [:field 87 nil]
    [:field 88 nil]
    [:field 81 nil]
    [:field 86 {:temporal-unit :default}]
    [:field 85 nil]
    [:field 106 {:source-field 82, :join-alias "PRODUCTS__via__PRODUCT_ID"}]],
   :joins
   [{:strategy :left-join,
     :alias "PRODUCTS__via__PRODUCT_ID",
     :condition [:= [:field 82 nil] [:field 105 {:join-alias "PRODUCTS__via__PRODUCT_ID"}]],
     :source-table 10,
     :fk-field-id 82}],
   :limit 1048575},
  :async? true},
 :data {:rows [], :cols []}}

2021-08-09 19:12:53,731 DEBUG middleware.log :: POST /api/card/723/query/xlsx 202 [ASYNC: completed] 1.1 s (17 DB calls) App DB connections: 0/10 Jetty threads: 3/50 (4 idle, 0 queued) (71 total active threads) Queries in flight: 1 (0 queued); h2 DB 4 connections: 0/1 (0 threads blocked)
```

</details>
<details><summary>5 metadata</summary>

```
{
  "dataset_query": {
    "database": 4,
    "query": {
      "source-table": 11,
      "limit": 1
    },
    "type": "query"
  },
  "display": "table",
  "displayIsLocked": true,
  "visualization_settings": {
    "table.columns": [
      {
        "name": "ID",
        "fieldRef": [
          "field",
          83,
          null
        ],
        "enabled": true
      },
      {
        "name": "USER_ID",
        "fieldRef": [
          "field",
          80,
          null
        ],
        "enabled": true
      },
      {
        "name": "PRODUCT_ID",
        "fieldRef": [
          "field",
          82,
          null
        ],
        "enabled": true
      },
      {
        "name": "SUBTOTAL",
        "fieldRef": [
          "field",
          84,
          null
        ],
        "enabled": true
      },
      {
        "name": "TAX",
        "fieldRef": [
          "field",
          87,
          null
        ],
        "enabled": true
      },
      {
        "name": "TOTAL",
        "fieldRef": [
          "field",
          88,
          null
        ],
        "enabled": true
      },
      {
        "name": "DISCOUNT",
        "fieldRef": [
          "field",
          81,
          null
        ],
        "enabled": true
      },
      {
        "name": "CREATED_AT",
        "fieldRef": [
          "field",
          86,
          {
            "temporal-unit": "default"
          }
        ],
        "enabled": true
      },
      {
        "name": "QUANTITY",
        "fieldRef": [
          "field",
          85,
          null
        ],
        "enabled": true
      },
      {
        "name": "ID_2",
        "fieldRef": [
          "field",
          105,
          {
            "join-alias": "Products"
          }
        ],
        "enabled": true
      },
      {
        "name": "EAN",
        "fieldRef": [
          "field",
          102,
          {
            "join-alias": "Products"
          }
        ],
        "enabled": true
      },
      {
        "name": "CATEGORY",
        "fieldRef": [
          "field",
          107,
          {
            "join-alias": "Products"
          }
        ],
        "enabled": true
      },
      {
        "name": "VENDOR",
        "fieldRef": [
          "field",
          109,
          {
            "join-alias": "Products"
          }
        ],
        "enabled": false
      },
      {
        "name": "PRICE",
        "fieldRef": [
          "field",
          104,
          {
            "join-alias": "Products"
          }
        ],
        "enabled": true
      },
      {
        "name": "RATING",
        "fieldRef": [
          "field",
          103,
          {
            "join-alias": "Products"
          }
        ],
        "enabled": true
      },
      {
        "name": "CREATED_AT_2",
        "fieldRef": [
          "field",
          108,
          {
            "temporal-unit": "default",
            "join-alias": "Products"
          }
        ],
        "enabled": true
      }
    ],
    "table.pivot_column": "CATEGORY",
    "table.cell_column": "SUBTOTAL",
    "column_settings": {
      "[\"ref\",[\"field\",108,null]]": {
        "date_style": "D MMMM, YYYY",
        "time_enabled": "seconds",
        "date_abbreviate": true
      },
      "[\"ref\",[\"field\",105,null]]": {
        "column_title": "Products222 → ID"
      },
      "[\"ref\",[\"field\",85,null]]": {
        "prefix": "Qty "
      },
      "[\"ref\",[\"field\",81,null]]": {
        "currency_style": "name"
      },
      "[\"ref\",[\"field\",86,null]]": {
        "column_title": "Created At2"
      }
    }
  },
  "original_card_id": 723
}
```

</details>

howonlee commented on src/metabase/query_processor/streaming/xlsx.clj:
```diff
@@ -22,61 +230,84 @@
+
+  Memoized so that it can be called within write-row! without re-running the logic to convert format settings
+  to format strings."
+  (memoize
```
making a note:
unmemoized I have samples from medium-size xlsx dl times of 2.12s 1.71s 1.60s 1.62s
memoized with same file it's 817ms 460ms 528ms 689 ms - so this deffo seems to be critical path

howonlee commented on src/metabase/query_processor/streaming/xlsx.clj:
```diff
@@ -1,17 +1,225 @@
+            (map #(str % "%") base-strings)
+
+            "scientific"
+            (map #(str % "E+0") base-strings)
```
this isn't actual scientific notation but I bet we punted on this somewhere. might add a ticket for figuring that out one day

noahmoss replied:
do you mean "scientific notation" != "E notation"?

on the frontend we display stuff in scientific notation, but this is the closest we can get in Excel

howonlee reviewed (commented):
everything looks copacetic, I don't have material stylistic comments, I tested on a bunch of weirdo inputs and it looks fine or messed up in a way that's not mb's fault, it's excel's, I will stamp as soon as you get the check warnings to go away

noahmoss commented:
@howonlee thanks for the review! what do you mean by check warnings?

dpsutton commented:
@noahmoss one that that surprised me is the shared visualization file has `model` in its namespace. `metabase.shared.models.visualization-settings` shouldn't have that as it has nothing "model"-y about it. Also, the require of `[cljs.js :as js]` is unnecessary for `js/JSON` in clojurescript. You can omit that require

flamber commented:
@noahmoss The "check warnings" - codecov - https://github.com/metabase/metabase/pull/17226/checks?check_run_id=3282728277

noahmoss committed:
Merge branch 'master' into export-formatting

noahmoss commented:
> one that that surprised me is the shared visualization file has `model` in its namespace

@jeff303 do you have context for why that file is in a `model` ns?

noahmoss committed:
fix typo

noahmoss committed:
remove js require in visualization_settings.cljc

noahmoss committed:
fix remapped fields for saved cards, and add tests for `export-column-order`

noahmoss commented:
> saved/public/embedded does not include the remapped field at all anymore

@flamber this should be fixed now

I also made the column ordering code ignore extra entries in `table.columns` that don't correspond to anything in `cols` which I think probably fixes that stacktrace you got.

noahmoss committed:
fix tests

noahmoss committed:
add a few tests that parse and assert on format strings in spreadsheet

noahmoss committed:
more tests for datetime types, and removed calls to deprecated .setCellType

noahmoss commented:
@howonlee I've removed the calls to `.setCellType` that I had copied over from Docjure since that method is apparently deprecated (https://poi.apache.org/apidocs/dev/org/apache/poi/ss/usermodel/Cell.html#setCellType-org.apache.poi.ss.usermodel.CellType-). From the experiments I've tried, I don't think this actually affects the risk of XLSX injection, but let me know if I'm mistaken.

flamber reviewed (approved):
Awesome. :+1: on all 5 points.

howonlee commented:
> @howonlee I've removed the calls to `.setCellType` that I had copied over from Docjure since that method is apparently deprecated (https://poi.apache.org/apidocs/dev/org/apache/poi/ss/usermodel/Cell.html#setCellType-org.apache.poi.ss.usermodel.CellType-). From the experiments I've tried, I don't think this actually affects the risk of XLSX injection, but let me know if I'm mistaken.

I can confirm no xlsx injection from our side

howonlee reviewed (approved):
check warnings fixed, everything looks copacetic still, stamp

noahmoss committed:
Merge branch 'master' into export-formatting

noahmoss committed:
empty commit to retrigger ci

noahmoss committed:
another emtpy commit to retrigger ci

noahmoss commented:
@flamber I'm not sure whether I should close #10879 now that I've merged this PR. The fixes for renamed, remapped, and hidden columns apply for all export formats, but datetime & number formatting only applies to XSLX.

That said, it's not clear which format settings should apply to CSVs/JSONs since the data might be destined for downstream processing rather than user consumption. Datetime formatting is probably still fair game, but maybe not number formatting.

Since this PR addresses maybe 75% of the original issue, should close that one and open a new one for CSV/JSON formatting?

ghiculescu commented:
> Datetime formatting is probably still fair game, but maybe not number formatting if it would require changing the data type from number to text.

If it helps: as a user, if I've chosen any formatting options for my column I'd expect to see the export formatted the same way. If the question is destined for downstream processing then I should keep that in build when building the question, it's not Metabase's responsibility to reformat the data.
